CMakeLists.txt

```diff
@@ -628,6 +629,28 @@ PUBLIC
 
 add_subdirectory(interface/basisu_c_binding)
 
+# Add astcenc subdirectory
+# Only one architecture is supported at once, if default (avx2) is not available,
+# user must define ISA_SSE41, ISA_SSE2 or ISA_NEON themselves
+
+set(ISA_AVX2 ON) # AVX2 is the default
+set(ASTC_LIB_NAME astcenc-avx2-static)
+
+if (${ISA_SSE41})
+  set(ASTC_LIB_NAME astcenc-sse4.1-static)
+endif()
+
+if (${ISA_SSE2})
+  set(ASTC_LIB_NAME astcenc-sse2-static)
+endif()
+
+if (${ISA_NEON})
+  set(ASTC_LIB_NAME astcenc-neon-static)
+endif()
+
+add_subdirectory(lib/astc-encoder)
+target_link_libraries(ktx PRIVATE ${ASTC_LIB_NAME})
+
```

Comment:
> AVX2 for Intel/AMD

Oops! I thought AVX2 was an ARM equivalent to SSE. I thought there was one that was not called Neon. I thought Neon was a specific ARM chip architecture.

Does this mean the libktx name also needs to change according to which instruction set is supported? I would prefer not. Which implementation should we choose for the installable Windows, Linux and macOS/Intel packages?

Currently for the BasisU encoder the library is always called libktx and the installers have the SSE version.

Reply:
As I just noted in a comment on build_android.sh I think a single ASTCENCODE_ISA option that takes the various values will be easier for people to understand.

Reply:
> I thought AVX2 was an ARM equivalent to SSE, I thought Neon was a specific ARM chip architecture.

[Neon](https://developer.arm.com/architectures/instruction-sets/simd-isas/neon) is the SIMD ISA (instruction set architecture) for armv8 and older architectures. The equivalent version i.e. the advanced version of NEON is called [SVE](https://developer.arm.com/tools-and-software/server-and-hpc/compile/arm-instruction-emulator/resources/tutorials/sve) for later armv8 and newer architectures. We don't have an SVE build option in `astcenc` yet.

> Does this mean the libktx name also needs to change according to which instruction set is supported?

It shouldn't. This only sets the ASTC shared library name. Its not going to be visible at install time.

> Which implementation should we choose for the installable Windows, Linux and macOS/Intel packages?

This is a very good question. Not sure what your strategy is for other things like this? One would usually go with the least common denominator but then you miss out on latest features when available. SSE4.1 for Intel and NEON for arm.

> Can the astcencoder be modified to similarly?

Probably. One could have some runtime paths for each SIMD architecture but making it build time option makes it easy to implement. Right now for NEON is easy, you either have it or you don't (will change with SVE). But for intel it gets complex to have different paths for each of AVX2, SSE4.1, SSE etc.

Its on our TODO list though. I submitted a request when I realized its shared libraries has different names.

Reply:
> the advanced version of NEON is called SVE

Ahh! That's what I'd been thinking of.

If there are several shared astcenc libraries with different names, how is it decided which one to run at program execution time or, in this case, when the libktx shared library is loaded?

The only other thing like this that we currently have is the SSE support in the Basis encoder. The installer has the version with SSE support as I think, perhaps mistakenly, that support for that is pretty much universal these days. We build and run tests on the non-SSE version too in CI to make sure it builds and runs.

Reply:
astcenc creates different libraries as well as different executables. You run what you want. `astcenc-avx2` etc. It is problematic and I have reported it. Pete is looking into fix that.

I think for astcenc we should also go with that which I am calling the least common denominator. SSE and NEON.

Reply:
what's the resolution on this one? I have had to make a few changes to this bit of cmake due to build issues. All the options are required by ASTCenc to build properly.

Reply:
I thought we'd decided to leave it as is because changes to the encoder handling of this are expected which we would wait for. We do need to avoid conflicts between the BASISU_SUPPORT_SSE setting and these build settings so a single library will work.